src/index/segment_reader.rs

```diff
@@ -213,83 +213,137 @@ impl SegmentReader {
         segment: &Segment,
         custom_bitset: Option<AliveBitSet>,
     ) -> crate::Result<SegmentReader> {
-        let start = std::time::Instant::now();
-        let segment_ref = SegmentRef::new(segment);
-        let termdict_composite_fut = executor.spawn(move || {
-            let termdict_file = segment_ref.open_read(SegmentComponent::Terms)?;
-            Ok(CompositeFile::open(&termdict_file)?)
-        })?;
-
-        let segment_ref = SegmentRef::new(segment);
-        let store_file_fut =
-            executor.spawn(move || Ok(segment_ref.open_read(SegmentComponent::Store)?))?;
-
-        crate::fail_point!("SegmentReader::open#middle");
+        use std::sync::Mutex;
 
-        let segment_ref = SegmentRef::new(segment);
-        let postings_composite_fut = executor.spawn(move || {
-            let postings_file = segment_ref.open_read(SegmentComponent::Postings)?;
-            Ok(CompositeFile::open(&postings_file)?)
-        })?;
+        let schema = segment.schema();
 
-        let segment_ref = SegmentRef::new(segment);
-        let positions_composite_fut = executor.spawn(move || {
-            if let Ok(positions_file) = segment_ref.open_read(SegmentComponent::Positions) {
-                Ok(CompositeFile::open(&positions_file)?)
-            } else {
-                Ok(CompositeFile::empty())
-            }
-        })?;
+        // Helper macro to spawn a task that stores result in a Mutex
+        macro_rules! spawn_load {
+            ($scope:expr, $result:expr, $body:expr) => {
+                $scope.spawn(|_| {
+                    *$result.lock().unwrap() = Some($body);
+                });
+            };
+        }
 
-        let schema = segment.schema();
+        // Use scope-based parallelism to avoid deadlock when called from within executor.map()
+        executor.install_and_scope(|| {
+            // Check if we have a thread pool available
+            if let Some(pool) = executor.thread_pool() {
+                // Use Rayon scope for work-stealing parallelism
+                let results = (
+                    Mutex::new(None), // termdict_composite
+                    Mutex::new(None), // store_file
+                    Mutex::new(None), // postings_composite
+                    Mutex::new(None), // positions_composite
+                    Mutex::new(None), // fast_fields_readers
+                    Mutex::new(None), // fieldnorm_readers
+                );
+
+                pool.scope(|scope| {
```

Comment:
Curious any reason not to prefer `pool.join`? The rayon docs seem to suggest that if we're spawning a fixed number of tasks, then join might be cheaper because it can allocate on the stack (https://docs.rs/rayon-core/1.13.0/rayon_core/fn.scope.html).